src/Autofac/Core/Resolving/InstanceLookup.cs

```diff
@@ -81,19 +83,43 @@ public object Execute()
             _executed = true;
 
             object? decoratorTarget = null;
-            object instance = ComponentRegistration.Sharing == InstanceSharing.None
-                ? Activate(Parameters, out decoratorTarget)
-                : _activationScope.GetOrCreateAndShare(ComponentRegistration.Id, () => Activate(Parameters, out decoratorTarget));
+
+            var sharing = _decoratorTargetComponent != null
+                ? _decoratorTargetComponent.Sharing
+                : ComponentRegistration.Sharing;
+
+            var resolveParameters = Parameters as Parameter[] ?? Parameters.ToArray();
+
+            _newInstance = _activationScope.Get(ComponentRegistration.Id);
+
+            if (_newInstance == null)
+            {
+                _newInstance = sharing == InstanceSharing.Shared
+                    ? _activationScope.Create(ComponentRegistration.Id, () => CreateInstance(Parameters))
+                    : CreateInstance(Parameters);
+            }
+
+            decoratorTarget = _newInstance;
+
+            _newInstance = InstanceDecorator.TryDecorateRegistration(
+                _service,
+                ComponentRegistration,
+                _newInstance,
+                _activationScope,
+                resolveParameters);
+
+            if (_newInstance != decoratorTarget)
+                ComponentRegistration.RaiseActivating(this, resolveParameters, ref _newInstance);
 
             var handler = InstanceLookupEnding;
             handler?.Invoke(this, new InstanceLookupEndingEventArgs(this, NewInstanceActivated));
 
             StartStartableComponent(decoratorTarget);
 
-            return instance;
+            return _newInstance;
         }
 
-        private void StartStartableComponent(object? instance)
+        private void StartStartableComponent(object instance)
```

Comment:
I think the `object?` was part of the nullable object support recently added. The warnings only show up in a few cases, though I don't remember what those cases are. If this thing can take in a null, I think the `object?` needs to stay in place.

Reply:
In the original code, Roslyn couldn't work out something always assigned to decoratorTarget in the Execute method, so the nullable marker was necessary. Looks like Roslyn can figure it out with the new method, so there's no need for the nullable marker in StartStartableComponent.